File: go/aggregate.go
/* Copyright (c) 2026 Richard Rodger and other contributors, MIT License */

package tabnasproto

// An aggregate option value: `option (foo) = { a: 1 };`.
//
// Port of ts/src/aggregate.ts, which explains protoc's rule in full. In
// short: protoc records the source text between the braces as the option's
// aggregate_value, every token, space and newline as written, with each run
// of comments replaced by the newlines and spaces that keep the next token
// on its line and column. A comment directly ahead of the closing brace
// leaves nothing. Columns are protoc's: a tab moves to the next multiple of
// 8, and every other byte is one column.

import (
	"regexp"
	"strings"

	tabnas "github.com/tabnas/parser/go"
)

const tabWidth = 8

// aggregateText is the text protoc 36 records for the aggregate whose
// braces are at src[open] and src[close].
//
// A Go string is bytes, so a byte offset is a column step already: a
// multi-byte character advances the column once per byte, as protoc's
// tokenizer does.
func aggregateText(src string, open, close int) string {
	line := 0
	col := 0
	step := func(i int) int {
		switch src[i] {
		case '\n':
			line++
			col = 0
		case '\t':
			col += tabWidth - col%tabWidth
		default:
			col++
		}
		return i + 1
	}

	// The column after the opening brace depends on everything ahead of it
	// on its line, because that decides where a later tab stops.
	for i := strings.LastIndexByte(src[:open], '\n') + 1; i <= open; {
		i = step(i)
	}
	line = 0

	isCommentStart := func(i int) bool {
		return src[i] == '/' && i+1 < len(src) && (src[i+1] == '/' || src[i+1] == '*')
	}

	var out strings.Builder
	from := open + 1
	i := from
	for i < close {
		c := src[i]
		if isCommentStart(i) {
			out.WriteString(src[from:i])
			gapLine, gapCol := line, col
			// Comments with nothing between them form one gap: protoc pads
			// from the token before the first to the token after the last.
			for i < close && isCommentStart(i) {
				if src[i+1] == '/' {
					for i < close && src[i] != '\n' {
						i = step(i)
					}
					if i < close {
						i = step(i)
					}
				} else {
					i = step(step(i))
					for i < close && !(src[i] == '*' && i+1 < len(src) && src[i+1] == '/') {
						i = step(i)
					}
					if i < close {
						i = step(step(i))
					}
				}
			}
			from = i
			if close <= i {
				break
			}
			if gapLine < line {
				out.WriteString(strings.Repeat("\n", line-gapLine))
				out.WriteString(strings.Repeat(" ", col))
			} else if gapCol < col {
				out.WriteString(strings.Repeat(" ", col-gapCol))
			}
			continue
		}
		if c == '"' || c == '\'' {
			// A string's text is copied as written; a `//` inside one is not
			// a comment. protoc ends an unterminated string at the newline.
			i = step(i)
			for i < close && src[i] != c && src[i] != '\n' {
				if src[i] == '\\' && i+1 < close && src[i+1] != '\n' {
					i = step(i)
				}
				i = step(i)
			}
			if i < close && src[i] == c {
				i = step(i)
			}
			continue
		}
		i = step(i)
	}
	if from < close {
		out.WriteString(src[from:close])
	}
	return out.String()
}

// recordAggregate is the after-close action Proto installs on the grammar's
// `constant` rule: on an aggregate value's CST node it sets "aggregate" to
// the text protoc records, read from the source between the value's braces.
// The node's "src" stays the tokens run together, as on every node. The
// opening brace is the rule's first token (O0) and the closing brace the
// last token consumed (V1); a node where either is not a brace is left
// without "aggregate". Port of recordAggregate in ts/src/aggregate.ts.
func recordAggregate(r *tabnas.Rule, ctx *tabnas.Context) {
	node, ok := r.Node.(map[string]any)
	if !ok || nrule(node) != "constant" {
		return
	}
	s, _ := node["src"].(string)
	if s == "" || s[0] != '{' || r.O0 == nil || ctx.V1 == nil || ctx.Lex == nil {
		return
	}
	src := ctx.Lex.Src
	open, close := r.O0.SI, ctx.V1.SI
	if !(0 <= open && open < close && close < len(src)) {
		return
	}
	if src[open] != '{' || src[close] != '}' {
		return
	}
	node["aggregate"] = aggregateText(src, open, close)
}

// ---- words inside an aggregate --------------------------------------------
//
// Port of the matcher in ts/src/aggregate.ts, which explains it in full.
// Text format has no keywords, so inside an aggregate value this matcher,
// which Proto runs ahead of the grammar's own, lexes as an identifier (#TX):
//
//   - a word the grammar spells as a keyword, always;
//   - `true`, `false`, `null`, `export` and `local` where they name a field:
//     followed by `:`, `{`, `<` or `[`;
//   - `[x.y]` or `[type.googleapis.com/x.Y]` followed by `:`, `{`, `<` or
//     `[`, the name between the brackets read as one, as text format reads
//     it.
//
// A `[` after a name opens a list of messages with the colon left out. A
// value followed by a bracketed name (`a: true [x.y]: 1`) is then read as an
// identifier too, which the grammar takes as a value: the entries and the
// text recorded are the same, and only that value's CST node differs.

// aggregateKeywords is every word the grammar spells as a literal, less
// `export` and `local`. TestAggregateKeywordsMatchTheGrammar keeps it in
// step with the grammar.
var aggregateKeywords = map[string]bool{
	"syntax": true, "import": true, "weak": true, "public": true, "package": true,
	"option": true, "message": true, "required": true, "optional": true,
	"repeated": true, "oneof": true, "map": true, "enum": true, "service": true,
	"rpc": true, "stream": true, "returns": true, "extend": true,
	"extensions": true, "reserved": true, "to": true, "max": true, "group": true,
	"edition": true,
}

// aggregatePunct is the grammar's punctuation, where the lexer ends a word.
const aggregatePunct = "{}[]:,;=()<>-.+"

func isWordStart(c byte) bool {
	return ('A' <= c && c <= 'Z') || ('a' <= c && c <= 'z') || c == '_'
}

func isDigit(c byte) bool { return '0' <= c && c <= '9' }

func isWordPart(c byte) bool { return isWordStart(c) || isDigit(c) }

func isLexSpace(c byte) bool { return c == ' ' || c == '\t' || c == '\r' || c == '\n' }

// endsWord reports whether the lexer ends a word at src[i].
func endsWord(src string, i int) bool {
	if i >= len(src) {
		return true
	}
	c := src[i]
	if isLexSpace(c) || strings.IndexByte(aggregatePunct, c) >= 0 || c == '#' {
		return true
	}
	return c == '/' && i+1 < len(src) && (src[i+1] == '/' || src[i+1] == '*')
}

// skipSpace returns the index of the first byte at or after i that is neither
// space nor inside a comment, or len(src). A comment ends where the lexer ends
// it: `#` and `//` at a CR or an LF, `/*` after the next `*/`.
func skipSpace(src string, i int) int {
	for i < len(src) {
		c := src[i]
		switch {
		case isLexSpace(c):
			i++
		case c == '#' || (c == '/' && i+1 < len(src) && src[i+1] == '/'):
			for i < len(src) && src[i] != '\n' && src[i] != '\r' {
				i++
			}
		case c == '/' && i+1 < len(src) && src[i+1] == '*':
			end := strings.Index(src[i+2:], "*/")
			if end < 0 {
				i = len(src)
			} else {
				i += 2 + end + 2
			}
		default:
			return i
		}
	}
	return i
}

// namesField reports whether a field name ends at i: is the next thing a
// `:`, `{`, `<` or `[`?
func namesField(src string, i int) bool {
	j := skipSpace(src, i)
	if j >= len(src) {
		return false
	}
	c := src[j]
	return c == ':' || c == '{' || c == '<' || c == '['
}

// typeNameRe is a type name as text format checks one.
var typeNameRe = regexp.MustCompile(`^[A-Za-z_][A-Za-z0-9_]*(\.[A-Za-z_][A-Za-z0-9_]*)*$`)

// bracketName reads `[x.y]` or `[type.googleapis.com/x.Y]` at open: the index
// after its `]` and the name with its spaces and comments left out. Port of
// bracketName in ts/src/aggregate.ts, which explains the two readings the
// name must pass: protoc's tokenizer's, which refuses a malformed number
// (numberEnd) and a decimal point with a digit after it directly behind an
// identifier; and text format's, which joins the pieces and wants a type
// name after the last `/` and a prefix that does not start with `/`. Only
// letters, digits, `_`, `.`, `/` and `-` are read, and the sign of a
// number's exponent, where text format also allows the URL characters
// `~!$&()*+,;=%` in a prefix.
func bracketName(src string, open int) (int, string, bool) {
	i := open + 1
	var name strings.Builder
	// identEnd is where the last identifier ended.
	identEnd := -1
	for {
		i = skipSpace(src, i)
		c := byteAt(src, i)
		switch {
		case isWordStart(c):
			from := i
			for isWordPart(byteAt(src, i)) {
				i++
			}
			name.WriteString(src[from:i])
			identEnd = i
			continue
		case isDigit(c) || (c == '.' && isDigit(byteAt(src, i+1))):
			if identEnd == i {
				return 0, "", false
			}
			end := numberEnd(src, i)
			if end < 0 {
				return 0, "", false
			}
			name.WriteString(src[i:end])
			i = end
			continue
		case c == '.' || c == '/' || c == '-':
			name.WriteByte(c)
			i++
			continue
		}
		if c != ']' {
			return 0, "", false
		}
		n := name.String()
		slash := strings.LastIndexByte(n, '/')
		if !typeNameRe.MatchString(n[slash+1:]) || strings.HasPrefix(n, "/") {
			return 0, "", false
		}
		return i + 1, "[" + n + "]", true
	}
}

// byteAt is src[i], or 0 past the end.
func byteAt(src string, i int) byte {
	if i < len(src) {
		return src[i]
	}
	return 0
}

// numberEnd is the index after the number protoc's tokenizer reads at at, a
// digit or a decimal point with a digit after it, or -1 where it refuses the
// number. Port of numberEnd in ts/src/aggregate.ts (protobuf's
// ConsumeNumber).
func numberEnd(src string, at int) int {
	dot, zero := src[at] == '.', src[at] == '0'
	i := at + 1
	switch c := byteAt(src, i); {
	case zero && (c == 'x' || c == 'X'):
		i++
		if !isHex(byteAt(src, i)) {
			return -1
		}
		for isHex(byteAt(src, i)) {
			i++
		}
	case zero && isDigit(c):
		for isOctal(byteAt(src, i)) {
			i++
		}
		if isDigit(byteAt(src, i)) {
			return -1
		}
	default:
		for isDigit(byteAt(src, i)) {
			i++
		}
		if !dot && byteAt(src, i) == '.' {
			i++
			for isDigit(byteAt(src, i)) {
				i++
			}
		}
		if e := byteAt(src, i); e == 'e' || e == 'E' {
			i++
			if s := byteAt(src, i); s == '-' || s == '+' {
				i++
			}
			if !isDigit(byteAt(src, i)) {
				return -1
			}
			for isDigit(byteAt(src, i)) {
				i++
			}
		}
	}
	if c := byteAt(src, i); isWordStart(c) || c == '.' {
		return -1
	}
	return i
}

func isOpenBrace(t *tabnas.Token) bool { return t != nil && t.Src == "{" }

// insideKey is the keep prop that marks a rule inside an aggregate value.
const insideKey = "protoAggregate"

// markAggregate is the before-open action Proto installs on every rule the
// grammar's `constant` rule pushes: a rule whose parent is an aggregate
// value's `constant`, the one whose first token is its `{`, is marked. The
// engine copies keep props to each rule pushed below a rule and to a rule
// that replaces one, so each rule inside the value carries the mark and no
// rule outside does. Port of markAggregate in ts/src/aggregate.ts.
func markAggregate(r *tabnas.Rule, _ *tabnas.Context) {
	p := r.Parent
	if p != nil && p != tabnas.NoRule && p.Name == "constant" && isOpenBrace(p.O0) {
		r.EnsureK()[insideKey] = true
	}
}

// pushedRules is the rules the grammar's `constant` rule pushes, which
// markAggregate is installed on.
func pushedRules(rs *tabnas.RuleSpec) []string {
	seen := map[string]bool{}
	var names []string
	for _, alt := range rs.OpenAlts() {
		if alt != nil && alt.P != "" && !seen[alt.P] {
			seen[alt.P] = true
			names = append(names, alt.P)
		}
	}
	return names
}

// inAggregate reports whether the lexer is inside an aggregate value. Every
// rule inside one carries the mark markAggregate sets, bar the value's
// `constant` itself: that is inside once its first token is the `{`, and
// while it is still choosing its alternative it peeks the tokens after the
// brace, with the brace already in the lookahead. Each test is a lookup, so
// the answer costs the same however deep the rule stack is.
func inAggregate(lex *tabnas.Lex, rule *tabnas.Rule) bool {
	if rule == nil || rule == tabnas.NoRule {
		return false
	}
	if inside, _ := rule.K[insideKey].(bool); inside {
		return true
	}
	if rule.Name != "constant" {
		return false
	}
	return isOpenBrace(rule.O0) ||
		(lex.Ctx != nil && len(lex.Ctx.T) > 0 && isOpenBrace(lex.Ctx.T[0]))
}

// aggregateWord is the lexer matcher: an identifier token, or nil to let the
// grammar's own matchers read the text.
func aggregateWord(lex *tabnas.Lex, rule *tabnas.Rule) *tabnas.Token {
	src := lex.Src
	pnt := lex.Cursor()
	at := pnt.SI
	if at >= len(src) {
		return nil
	}
	var end int
	var text string
	c := src[at]
	switch {
	case c == '[':
		e, name, ok := bracketName(src, at)
		if !ok || !namesField(src, e) {
			return nil
		}
		end, text = e, name
	case isWordStart(c):
		end = at + 1
		for end < len(src) && isWordPart(src[end]) {
			end++
		}
		if !endsWord(src, end) {
			return nil
		}
		text = src[at:end]
		lower := strings.ToLower(text)
		name := text == "true" || text == "false" || text == "null" ||
			lower == "export" || lower == "local"
		if !aggregateKeywords[lower] && !(name && namesField(src, end)) {
			return nil
		}
	default:
		return nil
	}
	if !inAggregate(lex, rule) {
		return nil
	}
	tkn := lex.Token("#TX", tabnas.TinTX, text, text)
	// A bracketed name may cross lines; keep the row and column the lexer
	// keeps (a newline starts a row at column 1).
	for i := at; i < end; i++ {
		if src[i] == '\n' {
			pnt.RI++
			pnt.CI = 1
		} else {
			pnt.CI++
		}
	}
	pnt.SI = end
	return tkn
}

// makeAggregateWord is the matcher's factory, for the engine's Lex.Match
// option.
func makeAggregateWord(_ *tabnas.LexConfig, _ *tabnas.Options) tabnas.LexMatcher {
	return aggregateWord
}

File: ts/src/aggregate.ts
/* Copyright (c) 2026 Richard Rodger and other contributors, MIT License */

// An aggregate option value: `option (foo) = { a: 1 };`.
//
// protoc's parser does not read the text between the braces. It records
// that text as `UninterpretedOption.aggregate_value`, and the text-format
// parser reads it later, once the option's type is known. Since protobuf
// v36 the recorded text keeps its layout (`ParseUninterpretedBlock` in
// src/google/protobuf/compiler/parser.cc). protoc walks the tokens between
// the braces with whitespace and newlines reported as tokens of their own,
// and appends each token's text as written. Comments are not tokens, and
// text format does not read `//` or `/* */`, so before each token protoc
// pads out whatever lay between it and the token before it, the opening
// brace for the first:
//
// - a token that starts on a later line gets one newline per line crossed,
//   then as many spaces as its column;
// - a token that starts further along the same line gets the difference
//   in spaces.
//
// Whitespace is a token, so the only thing ever padded out is a run of
// comments. A comment within one line becomes spaces of its width. A
// block comment that crosses lines, or a line comment, which protoc ends
// AFTER its newline, becomes its newlines followed by spaces up to the
// column where it ends. protoc stops at the closing brace before padding
// for it, so a comment directly ahead of that brace leaves nothing.
//
// Columns are counted as protoc's tokenizer counts them: a tab moves to
// the next multiple of 8, and every other byte of UTF-8 is one column.

const TAB = 8

// The text protoc 36 records for the aggregate whose braces are at
// `src[open]` and `src[close]`.
function aggregateText(src: string, open: number, close: number): string {
  let line = 0
  let col = 0

  // Move past `src[i]`, keeping `line` and `col` protoc's way. Returns the
  // index after it: a surrogate pair is one character and four bytes.
  const step = (i: number): number => {
    const c = src.charCodeAt(i)
    if (10 === c) {
      line++
      col = 0
    } else if (9 === c) {
      col += TAB - (col % TAB)
    } else if (c < 0x80) {
      col += 1
    } else if (c < 0x800) {
      col += 2
    } else if (0xd800 <= c && c < 0xdc00 && i + 1 < src.length) {
      const d = src.charCodeAt(i + 1)
      if (0xdc00 <= d && d < 0xe000) {
        col += 4
        return i + 2
      }
      col += 3
    } else {
      col += 3
    }
    return i + 1
  }

  // The column after the opening brace depends on everything ahead of it
  // on its line, because that decides where a later tab stops.
  for (let i = src.lastIndexOf('\n', open) + 1; i <= open;) i = step(i)
  line = 0

  const isCommentStart = (i: number): boolean =>
    '/' === src[i] && ('/' === src[i + 1] || '*' === src[i + 1])

  let out = ''
  let from = open + 1
  let i = from
  while (i < close) {
    const c = src[i]
    if (isCommentStart(i)) {
      out += src.slice(from, i)
      const gapLine = line
      const gapCol = col
      // Comments with nothing between them form one gap: protoc pads from
      // the token before the first to the token after the last.
      while (i < close && isCommentStart(i)) {
        if ('/' === src[i + 1]) {
          while (i < close && '\n' !== src[i]) i = step(i)
          if (i < close) i = step(i)
        } else {
          i = step(step(i))
          while (i < close && !('*' === src[i] && '/' === src[i + 1])) i = step(i)
          if (i < close) i = step(step(i))
        }
      }
      from = i
      if (close <= i) break
      if (gapLine < line) out += '\n'.repeat(line - gapLine) + ' '.repeat(col)
      else if (gapCol < col) out += ' '.repeat(col - gapCol)
      continue
    }
    if ('"' === c || "'" === c) {
      // A string's text is copied as written; a `//` inside one is not a
      // comment. protoc ends an unterminated string at the newline.
      i = step(i)
      while (i < close && c !== src[i] && '\n' !== src[i]) {
        if ('\\' === src[i] && i + 1 < close && '\n' !== src[i + 1]) i = step(i)
        i = step(i)
      }
      if (i < close && c === src[i]) i = step(i)
      continue
    }
    i = step(i)
  }
  return out + src.slice(from, close)
}

type AggregateNode = { rule?: unknown; src?: unknown; aggregate?: string }

// After-close action for the grammar's `constant` rule, which the `Proto`
// plugin installs: on an aggregate value's CST node it sets `aggregate` to
// the text protoc records, read from the source between the value's
// braces. The node's `src` stays the tokens run together, as on every
// node. The opening brace is the rule's first token (`o0`) and the closing
// brace the last token consumed (`v1`); a node where either is not a brace
// is left without `aggregate`.
function recordAggregate(rule: any, ctx: any): void {
  const node: AggregateNode | undefined = rule?.node
  if (!node || 'constant' !== node.rule) return
  if ('string' !== typeof node.src || '{' !== node.src[0]) return
  const src: string = ctx.src()
  const open: number = rule.o0?.sI
  const close: number = ctx.v1?.sI
  if (!(0 <= open && open < close && close < src.length)) return
  if ('{' !== src[open] || '}' !== src[close]) return
  node.aggregate = aggregateText(src, open, close)
}

// ---- words inside an aggregate --------------------------------------------
//
// Text format has no keywords: `message`, `optional` and `max` are field
// names and enum values like any other word. The grammar spells its own
// keywords as whole-word literals, and the lexer takes one wherever it
// occurs, so inside an aggregate every such word would be refused where
// an identifier belongs. This matcher, which the `Proto` plugin runs ahead
// of the grammar's own, lexes those words as identifiers (#TX) inside an
// aggregate value and nowhere else, as text format reads them:
//
// - a word the grammar spells as a keyword, always;
// - `true`, `false`, `null` (values to the lexer) and `export`, `local`
//   (which stay keywords in a value such as `a: local.x`, as they always
//   have), where they name a field: followed by `:`, `{`, `<` or `[`;
// - a field named by an extension or an Any type URL, `[x.y]` or
//   `[type.googleapis.com/x.Y]`, followed by `:`, `{`, `<` or `[`. Text
//   format reads the text between those brackets as one name, and so does
//   this; anywhere else a `[` opens a list.
//
// A `[` after a name opens a list of messages with the colon left out, as
// text format allows (`[x.y] [ { a: 1 } ]`). What comes before that `[`
// may instead be a value, with a bracketed name after it: `a: true [x.y]:
// 1`, or `a: [x] [y]: 1`. Such a value is read as an identifier too. The
// grammar takes an identifier as a value, so the entries are the same and
// so is the text recorded for the aggregate; only that value's CST node is
// an identifier's rather than a list's or a literal's.
//
// A word glued to a character that would carry the lexer's text on
// (`message/x`) is left alone: the identifier made here is always the one
// the lexer would make if the word were not a keyword.

// Every word the grammar spells as a literal, less `export` and `local`.
// test/ keeps this in step with the grammar.
const KEYWORDS = new Set([
  'syntax', 'import', 'weak', 'public', 'package', 'option', 'message',
  'required', 'optional', 'repeated', 'oneof', 'map', 'enum', 'service',
  'rpc', 'stream', 'returns', 'extend', 'extensions', 'reserved', 'to',
  'max', 'group', 'edition',
])

// Where the lexer ends a word: whitespace, the grammar's punctuation, a
// comment, or the end of the source.
const PUNCT = '{}[]:,;=()<>-.+'

const isWordStart = (c: number): boolean =>
  (0x41 <= c && c <= 0x5a) || (0x61 <= c && c <= 0x7a) || 0x5f === c
const isDigit = (c: number): boolean => 0x30 <= c && c <= 0x39
const isWordPart = (c: number): boolean => isWordStart(c) || isDigit(c)
const isSpace = (c: string): boolean => ' ' === c || '\t' === c || '\r' === c || '\n' === c

function endsWord(src: string, i: number): boolean {
  if (src.length <= i) return true
  const c = src[i]
  if (isSpace(c) || PUNCT.includes(c) || '#' === c) return true
  return '/' === c && ('/' === src[i + 1] || '*' === src[i + 1])
}

// The index of the first character at or after `i` that is neither space
// nor inside a comment, or src.length. A comment ends where the lexer ends
// it: `#` and `//` at a CR or an LF, `/*` after the next `*/`.
function skipSpace(src: string, i: number): number {
  while (i < src.length) {
    const c = src[i]
    if (isSpace(c)) i++
    else if ('#' === c || ('/' === c && '/' === src[i + 1])) {
      while (i < src.length && '\n' !== src[i] && '\r' !== src[i]) i++
    } else if ('/' === c && '*' === src[i + 1]) {
      const end = src.indexOf('*/', i + 2)
      i = -1 === end ? src.length : end + 2
    } else break
  }
  return i
}

// Does a field name end at `i`: is the next thing a `:`, `{`, `<` or `[`?
const namesField = (src: string, i: number): boolean => {
  const c = src[skipSpace(src, i)]
  return ':' === c || '{' === c || '<' === c || '[' === c
}

// A type name as text format checks one: identifiers joined by dots.
const TYPE_NAME = /^[A-Za-z_][A-Za-z0-9_]*(\.[A-Za-z_][A-Za-z0-9_]*)*$/

// `[x.y]` or `[type.googleapis.com/x.Y]` at `open`: the index after its
// `]` and the name with its spaces and comments left out, or null.
//
// The name must pass the two readings protoc gives it. protoc's own
// tokenizer reads the text first, as it reads everything between an
// aggregate's braces, and refuses the file where a number runs into a
// letter or a `_` (`[1p.example/x.Y]`), a number has a second decimal
// point (`[127.0.0.1/x.Y]`), an octal or hex number is malformed
// (`[009/x.Y]`), or a decimal point with a digit after it follows an
// identifier directly (`[a.2/x.Y]`). Text format
// (`ConsumeAnyTypeUrlOrFullTypeName` in protobuf's text_format.cc) then
// joins what the brackets hold, spaces and comments left out, so
// `[x.y 2]` names `x.y2`, and wants a type name after the last `/`, if
// there is one, and a prefix before it that does not start with `/`.
//
// Only letters, digits, `_`, `.`, `/` and `-` are read, and the sign of a
// number's exponent. Text format also allows the URL characters
// `~!$&()*+,;=%` in a prefix; a name holding one is not read as a name
// here, and the aggregate is refused.
function bracketName(src: string, open: number): { end: number; name: string } | null {
  let i = open + 1
  let name = ''
  // Where the last identifier ended: a decimal point there, with a digit
  // after it, is refused (protoc: "Need space between identifier and
  // decimal point").
  let identEnd = -1
  for (;;) {
    i = skipSpace(src, i)
    const c = src.charCodeAt(i)
    if (isWordStart(c)) {
      const from = i
      while (isWordPart(src.charCodeAt(i))) i++
      name += src.slice(from, i)
      identEnd = i
    } else if (isDigit(c) || (0x2e === c && isDigit(src.charCodeAt(i + 1)))) {
      if (identEnd === i) return null
      const end = numberEnd(src, i)
      if (end < 0) return null
      name += src.slice(i, end)
      i = end
    } else if (0x2e === c || 0x2f === c || 0x2d === c) {
      name += src[i++]
    } else {
      break
    }
  }
  if (']' !== src[i]) return null
  const slash = name.lastIndexOf('/')
  if (!TYPE_NAME.test(name.slice(slash + 1)) || 0 === name.indexOf('/')) return null
  return { end: i + 1, name: '[' + name + ']' }
}

const isHex = (c: number): boolean =>
  isDigit(c) || (0x41 <= c && c <= 0x46) || (0x61 <= c && c <= 0x66)
const isOctal = (c: number): boolean => 0x30 <= c && c <= 0x37

// The index after the number protoc's tokenizer reads at `at`, a digit or
// a decimal point with a digit after it, or -1 where it refuses the
// number. This is `ConsumeNumber` in protobuf's io/tokenizer.cc.
function numberEnd(src: string, at: number): number {
  const code = (k: number): number => src.charCodeAt(k)
  const dot = 0x2e === code(at)
  const zero = 0x30 === code(at)
  let i = at + 1
  if (zero && (0x78 === code(i) || 0x58 === code(i))) {
    // Hexadecimal: `0x` and at least one hex digit.
    i++
    if (!isHex(code(i))) return -1
    while (isHex(code(i))) i++
  } else if (zero && isDigit(code(i))) {
    // Octal: a leading zero, and octal digits only.
    while (isOctal(code(i))) i++
    if (isDigit(code(i))) return -1
  } else {
    while (isDigit(code(i))) i++
    if (!dot && 0x2e === code(i)) {
      i++
      while (isDigit(code(i))) i++
    }
    if (0x65 === code(i) || 0x45 === code(i)) {
      // An exponent: `e`, a sign if any, and at least one digit.
      i++
      if (0x2d === code(i) || 0x2b === code(i)) i++
      if (!isDigit(code(i))) return -1
      while (isDigit(code(i))) i++
    }
  }
  // Neither a letter nor another decimal point may follow.
  return isWordStart(code(i)) || 0x2e === code(i) ? -1 : i
}

const isOpenBrace = (t: any): boolean => null != t && '{' === t.src

// The keep prop that marks a rule inside an aggregate value.
const INSIDE = 'protoAggregate'

// Before-open action, which the `Proto` plugin installs on every rule the
// grammar's `constant` rule pushes: a rule whose parent is an aggregate
// value's `constant`, the one whose first token is its `{`, is marked. The
// mark is a keep prop, and the engine copies keep props to each rule
// pushed below a rule and to a rule that replaces one, in every runtime,
// so each rule inside the value carries it and no rule outside does. It
// is set on the rules the `constant` pushes rather than on the `constant`
// itself because the engine copies them at the push, which comes before
// an after-open action runs.
function markAggregate(rule: any): void {
  const parent = rule?.parent
  if ('constant' === parent?.name && isOpenBrace(parent.o0)) rule.k[INSIDE] = true
}

// The rules the grammar's `constant` rule pushes, which markAggregate is
// installed on.
function pushedRules(constant: any): string[] {
  const names = (constant?.def?.open || []).map((alt: any) => alt?.p)
  return [...new Set<string>(names.filter((p: any) => 'string' === typeof p))]
}

// Is the lexer inside an aggregate value? Every rule inside one carries
// the mark markAggregate sets, bar the value's `constant` itself: that is
// inside once its first token is the `{`, and while it is still choosing
// its alternative it peeks the tokens after the brace, with the brace
// already in the lookahead. Each test is a lookup, so the answer costs the
// same however deep the rule stack is.
function inAggregate(lex: any, rule: any): boolean {
  const keep = 'function' === typeof rule?.rawk ? rule.rawk() : rule?.k
  if (true === keep?.[INSIDE]) return true
  return 'constant' === rule?.name && (isOpenBrace(rule.o0) || isOpenBrace(lex.ctx?.t?.[0]))
}

// The lexer matcher. Returns an identifier token, or undefined to let the
// grammar's own matchers read the text.
function aggregateWord(lex: any, rule: any): any {
  const src: string = lex.src
  const pnt = lex.pnt
  const at: number = pnt.sI
  const c = src.charCodeAt(at)
  let end: number
  let text: string
  if (0x5b === c) {
    const bn = bracketName(src, at)
    if (null == bn || !namesField(src, bn.end)) return undefined
    end = bn.end
    text = bn.name
  } else if (isWordStart(c)) {
    end = at + 1
    while (end < src.length && isWordPart(src.charCodeAt(end))) end++
    if (!endsWord(src, end)) return undefined
    text = src.slice(at, end)
    const lower = text.toLowerCase()
    const name = 'true' === text || 'false' === text || 'null' === text ||
      'export' === lower || 'local' === lower
    if (!KEYWORDS.has(lower) && !(name && namesField(src, end))) return undefined
  } else {
    return undefined
  }
  if (!inAggregate(lex, rule)) return undefined
  const tkn = lex.token('#TX', text, text, pnt, undefined, undefined, end - at)
  // A bracketed name may cross lines; keep the row and column the lexer
  // keeps (a newline starts a row at column 1).
  for (let i = at; i < end; i++) {
    if ('\n' === src[i]) {
      pnt.rI++
      pnt.cI = 1
    } else {
      pnt.cI++
    }
  }
  pnt.sI = end
  return tkn
}

// The matcher's factory, for the engine's `lex.match` option.
const makeAggregateWord = () => aggregateWord

export {
  aggregateText, recordAggregate, makeAggregateWord, markAggregate, pushedRules,
  inAggregate, skipSpace, KEYWORDS,
}
